### PR TITLE
feat(losses): add Dice loss implementation

### PR DESCRIPTION
**Pull request by lpizzinidev**
Adds Dice class/function implementation to losses.

### CODE REVIEW

File: keras/losses/losses.py [hunk 1]
```diff
@@ -1934,3 +1934,76 @@ def ctc(y_true, y_pred):
...
+    def get_config(self):
+        return {
+            "name": self.name,
+            "reduction": self.reduction,
+            "smooth": self.smooth,
+        }
+
+
+@keras_export("keras.losses.dice")
+def dice(y_true, y_pred, smooth=1e-6):
```
Comment (fchollet) [inline keras/losses/losses.py:1981]: You can remove the `smooth` argument.

File: keras/losses/losses.py [hunk 2]
```diff
@@ -1934,3 +1934,76 @@ def ctc(y_true, y_pred):
...
+    """
+    y_true = ops.cast(y_true, dtype="float32")
+    y_pred = ops.cast(y_pred, dtype="float32")
+
+    inputs = ops.reshape(y_true, [-1])
+    targets = ops.reshape(y_pred, [-1])
+
+    intersection = ops.sum(ops.dot(inputs, targets))
+    dice = ops.divide(
+        2.0 * intersection + smooth, ops.sum(y_true) + ops.sum(y_pred) + smooth
```
Comment (fchollet) [inline keras/losses/losses.py:2006]: Instead `smooth`, use `backend.epsilon()`. Only use it for the denominator.

File: keras/losses/losses.py [hunk 3]
```diff
@@ -1934,3 +1934,76 @@ def ctc(y_true, y_pred):
...
+    Args:
+        y_true: tensor of true targets.
+        y_pred: tensor of predicted targets.
+        smooth: coefficient to avoid division by zero when both `y_true`
+            and `y_pred` do not contain any foreground pixels.
+
+    Returns:
+        Dice loss value.
+    """
+    y_true = ops.cast(y_true, dtype="float32")
```
Comment (fchollet) [inline keras/losses/losses.py:1998]: There's no need to force the use of `float32`, you can just use `ops.convert_to_tensor(y_true)`

File: keras/losses/losses.py [hunk 4]
```diff
@@ -1934,3 +1934,76 @@ def ctc(y_true, y_pred):
...
+    y_pred = ops.cast(y_pred, dtype="float32")
```
Comment (fchollet) [inline keras/losses/losses.py:1999]: Same here.

```
y_pred = ops.convert_to_tensor(y_pred)
y_true = ops.cast(y_true, y_pred.dtype)
```

**Review by fchollet — COMMENTED** [review]
Thank you for the PR!

**Comment by lpizzinidev** [comment]
@fchollet Thanks for reviewing 👍

**Comment by innat** [comment]
Shouldn't it be in keras-cv? 

https://github.com/keras-team/keras-cv/pull/371
https://github.com/keras-team/keras-cv/pull/968

File: keras/losses/losses.py [hunk 5]
```diff
@@ -1934,3 +1934,69 @@ def ctc(y_true, y_pred):
...
+    Returns:
+        Dice loss value.
+    """
+    y_true = ops.convert_to_tensor(y_true)
+    y_pred = ops.convert_to_tensor(y_pred)
+
+    inputs = ops.reshape(y_true, [-1])
+    targets = ops.reshape(y_pred, [-1])
+
+    intersection = ops.sum(ops.dot(inputs, targets))
```
Comment (fchollet) [inline keras/losses/losses.py:1996]: Easier to replace `dot` with `*` here (doesn't change numerics)

File: keras/losses/losses.py [hunk 6]
```diff
@@ -1934,3 +1934,69 @@ def ctc(y_true, y_pred):
...
+
+    Args:
+        y_true: tensor of true targets.
+        y_pred: tensor of predicted targets.
+
+    Returns:
+        Dice loss value.
+    """
+    y_true = ops.convert_to_tensor(y_true)
+    y_pred = ops.convert_to_tensor(y_pred)
```
Comment (fchollet) [inline keras/losses/losses.py:1991]: Do:

```
y_pred = ops.convert_to_tensor(y_pred)
y_true = ops.cast(y_true, y_pred.dtype)
```

**Review by fchollet — APPROVED** [review]
LGTM, thank you!

File: keras/losses/losses.py (hunk 6 above)
Comment (lpizzinidev) [reply keras/losses/losses.py:1990]: Sorry, I misunderstood the review.
What's the difference in the two approaches?
